### PR TITLE
[SPARK-32025][SQL] Csv schema inference problems with different types in the same column

### PR DESCRIPTION
**Pull request by planga82**
### What changes were proposed in this pull request?

This pull request fixes a bug present in the csv type inference.
We have problems when we have different types in the same column.

**Previously:** 
```
$ cat /example/f1.csv
col1
43200000
true

spark.read.csv(path="file:///example/*.csv", header=True, inferSchema=True).show()
+----+
|col1|
+----+
|null|
|true|
+----+

root
 |-- col1: boolean (nullable = true)
```
**Now**
```
spark.read.csv(path="file:///example/*.csv", header=True, inferSchema=True).show()
+-------------+
|col1          |
+-------------+
|43200000 |
|true           |
+-------------+

root                                                                            
 |-- col1: string (nullable = true)
```

Previously the hierarchy of type inference is the following:

> IntegerType
> > LongType
> > > DecimalType
> > > > DoubleType
> > > > > TimestampType
> > > > > > BooleanType
> > > > > > > StringType

So, when, for example, we have integers in one column, and the last element is a boolean, all the column is inferred as a boolean column incorrectly and all the number are shown as null when you see the data

We need the following hierarchy. When we have different numeric types in the column it will be resolved correctly. And when we have other different types it will be resolved as a String type column
> IntegerType
> > LongType
> > > DecimalType
> > > > DoubleType
> > > > > StringType

> TimestampType
> > StringType

> BooleanType
> > StringType

> StringType

### Why are the changes needed?

Fix the bug explained

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test and manual tests

### CODE REVIEW

**Comment by planga82** [comment]
CC @HyukjinKwon @cloud-fan @MaxGekk

**Review by MaxGekk — COMMENTED** [review]
If StringType is not the last type in any type inference chains, this can break existing users apps, I guess. Can't it?

**Comment by planga82** [comment]
@MaxGekk Yes, the last type in inference is StringType in all inference chains. I have change the description to show it better.
The principal difference is than now we have different chains instead of one

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -103,7 +103,16 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
       typeSoFar
     } else {
       typeSoFar match {
-        case NullType => tryParseInteger(field)
+        case NullType =>
+          tryParseInteger(field) match {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:107]: Let's don't have a separate rule alone here. The problem here seems like when the type is inferred from single partition alone, and the compatible types are not respected. You might have to call `compatibleType` from my cursory look.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:107]: But the problem is that that's going to degrade the performance, and single partition case is arguably rare. Let's make sure we don't have the performance penalty in general cases.

**Comment by cloud-fan** [comment]
Does JSON source have the same problem?

**Comment by planga82** [comment]
@HyukjinKwon It's exactly what you say, it only happens when the incompatibility is inside one partition. I will change the PR to use compatibleType, and I will make some performance test. Thanks for your help!

@cloud-fan I tested the same situation with json and it works fine, we don't have problems there

**Comment by planga82** [comment]
I have done some performance tests in my local machine
- File: 999999999 lines integers / 10 GB 
- Resources: 4 Cores

Result without changes:
- 21m10,114s
- 20m18,626s
- 19m46,158s

Results after changes
- 19m46,388s
- 22m10,784s
- 21m15,937s

It seems that we don't have a very significant impact but the tests in local are not the best way to be sure

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala [hunk 2]
```diff
@@ -82,7 +82,8 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
   def inferRowType(rowSoFar: Array[DataType], next: Array[String]): Array[DataType] = {
     var i = 0
     while (i < math.min(rowSoFar.length, next.length)) {  // May have columns on right missing.
-      rowSoFar(i) = inferField(rowSoFar(i), next(i))
+      val typeElemInfer = inferField(rowSoFar(i), next(i))
+      rowSoFar(i) = compatibleType(rowSoFar(i), typeElemInfer).getOrElse(StringType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:86]: shall we do this inside `inferField`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 3]
```diff
@@ -2341,6 +2342,17 @@ abstract class CSVSuite extends QueryTest with SharedSparkSession with TestCsvDa
       checkAnswer(csv, Row(null))
     }
   }
+
+  test("Test mixed types") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:2346]: nit: `SPARK-32025: infer the schema from mixed-type values`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala (hunk 2 above)
Comment (planga82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala]: Ok it's more intuitive and inferField is only used here and in the tests

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala [hunk 4]
```diff
@@ -102,7 +102,7 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
     if (field == null || field.isEmpty || field == options.nullValue) {
       typeSoFar
     } else {
-      typeSoFar match {
+      val typeElemInfer = typeSoFar match {
         case NullType => tryParseInteger(field)
         case IntegerType => tryParseInteger(field)
         case LongType => tryParseLong(field)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:108]: Seems like now you can just do `tryParseDecimal(field)` alone at decimal case below.

File: sql/core/src/test/resources/test-data/mixed-types1.csv [hunk 5]
```diff
@@ -0,0 +1,4 @@
+col_mixed_types
+2012
+1997
+True
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/test-data/mixed-types1.csv:4]: Let's don't bother to create new data file here. You write the test codes as below instead.

```scala
withTempPath { path =>
  Seq("col_mixed_types", "2012", "1997", "True").toDS.write.text(path.getCanonicalPath)
  val df = spark.read.format("csv")
    .option("header", "true")
    .option("inferSchema", "true")
    .load(path.getCanonicalPath)

  assert(data.schema.last == StructField("col_mixed_types", StringType, true))
}
```

Comment (planga82) [reply sql/core/src/test/resources/test-data/mixed-types1.csv:4]: Ok, It's more clear to see the data in the test.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
ok to test

**Review by HyukjinKwon — APPROVED** [review]

**Comment by planga82** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by gatorsmile** [comment]
Is that possible we can document the type inference rule? something like the traditional database like https://www.ibm.com/support/knowledgecenter/SSEPGG_10.1.0/com.ibm.db2.luw.sql.ref.doc/doc/r0008477.html

cc @huaxingao @planga82

**Comment by huaxingao** [comment]
@gatorsmile I will take a look.

**Comment by planga82** [comment]
Thanks @huaxingao  If you can't do it for any reason tell me.

**Pull request closed without merging**